src/services/WorkPeriodService.js

```diff
@@ -0,0 +1,500 @@
+    workPeriod.endDate = date
+  } else if (!workPeriod.startDate && workPeriod.endDate) {
+    const date = new Date(workPeriod.endDate)
+    date.setDate(date.getDate() - 6)
+    workPeriod.startDate = date
+  }
+
+  const resourceBooking = await helper.ensureResourceBookingById(workPeriod.resourceBookingId) // ensure resource booking exists
+  workPeriod.projectId = resourceBooking.projectId
+
+  const user = await helper.ensureUserById(resourceBooking.userId) // ensure user exists
+  workPeriod.userHandle = user.handle
+
+  workPeriod.id = uuid.v4()
+  workPeriod.createdBy = await helper.getUserId(currentUser.userId)
+
+  let created = null
+  try {
+    created = await WorkPeriod.create(workPeriod)
+  } catch (err) {
+    if (!_.isUndefined(err.original)) {
+      throw new errors.BadRequestError(err.original.detail)
+    } else {
+      throw err
+    }
+  }
+
+  await helper.postEvent(config.TAAS_WORK_PERIOD_CREATE_TOPIC, created.toJSON())
+  return created.dataValues
+}
+
+createWorkPeriod.schema = Joi.object().keys({
+  currentUser: Joi.object().required(),
+  workPeriod: Joi.object().keys({
+    resourceBookingId: Joi.string().uuid().required(),
+    startDate: Joi.workPeriodStartDate(),
+    endDate: Joi.workPeriodEndDate(),
+    daysWorked: Joi.number().integer().min(0).allow(null),
+    memberRate: Joi.number().allow(null),
+    customerRate: Joi.number().allow(null),
+    paymentStatus: Joi.paymentStatus().required()
+  }).required()
+}).required()
+
+/**
+  * Update workPeriod
+  * @param {Object} currentUser the user who perform this operation
+  * @param {String} id the workPeriod id
+  * @param {Object} data the data to be updated
+  * @returns {Object} the updated workPeriod
+  */
+async function updateWorkPeriod (currentUser, id, data) {
+  // check permission
+  await _checkUserPermissionForWriteWorkPeriod(currentUser)
+
+  const workPeriod = await WorkPeriod.findById(id)
+  const oldValue = workPeriod.toJSON()
+
+  // if resourceBookingId is provided then update projectId and userHandle
+  if (data.resourceBookingId) {
+    const resourceBooking = await helper.ensureResourceBookingById(data.resourceBookingId) // ensure resource booking exists
+    data.projectId = resourceBooking.projectId
+
+    const user = await helper.ensureUserById(resourceBooking.userId) // ensure user exists
+    data.userHandle = user.handle
+  }
+  // If one of the dates are missing then auto-calculate it
+  if (data.startDate && !data.endDate) {
+    const date = new Date(data.startDate)
+    date.setDate(date.getDate() + 6)
+    data.endDate = date
+  } else if (!data.startDate && data.endDate) {
+    const date = new Date(data.endDate)
+    date.setDate(date.getDate() - 6)
+    data.startDate = date
+  }
+  // change the date format to match with database model
+  if (data.startDate && data.endDate) {
+    data.startDate = moment(data.startDate).format('YYYY-MM-DD')
+    data.endDate = moment(data.endDate).format('YYYY-MM-DD')
+  }
```

Comment:
@eisbilir we have the logic `change the date format to match with database model` during the update, but not during creating. I wonder what is the correct way: can we remove this logic during update if it's not needed, or should we add the same logic during creating if its needed?

Reply:
We define `startDate` and `endDate` formats in `Joi.date().format('YYYY-MM-DD')`
This works as expected, Joi validates the date formats as we want, without timestamps.
But the problem is in here: `src/common/logger.js#149` --> `const normalized = Joi.attempt(value, method.schema)`
After joi validates our dates which is in `'YYYY-MM-DD'` format with above command; it returns dates localized with timestamps. So normalized variable becomes `2021-04-04T12:04:09.492Z`. I believe this is a bug.

This is why I converted it to desired format back.
Now, I tried a different logic and it seems that we don't need that conversion anymore because sequelize will convert them back to `'YYYY-MM-DD'`. But we also have remove the line #263 `const result = _.assign(workPeriod.dataValues, data)` and return `updated.dataValues` directly. With this way we won't need that conversion anymore.